### PR TITLE
framework/iotbus: add system I/O interrupt handling process

### PR DESCRIPTION
**Pull request by an4967**
- add gpio falling/rising interrupt handler in iotbus
- move iotbus_irq initialize from os_start to os_bringup

### CODE REVIEW

**Comment by seinfra** [comment]
Target : [4a5ed3295d3a895b79cc076b19b7d7721076b5aa] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [4a5ed3295d3a895b79cc076b19b7d7721076b5aa] - Code Rule Check Result:
framework/src/iotbus/iotbus_gpio.c:420: ERROR: [IDT_M_TAB] code indent should use tabs where possible
framework/src/iotbus/iotbus_gpio.c:420: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
framework/src/iotbus/iotbus_gpio.c:421: ERROR: [IDT_M_TAB] code indent should use tabs where possible
framework/src/iotbus/iotbus_gpio.c:421: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
framework/src/iotbus/iotbus_gpio.c:422: ERROR: [IDT_M_TAB] code indent should use tabs where possible
framework/src/iotbus/iotbus_gpio.c:422: ERROR: [IDT_M_TAB] please, no spaces at the start of a line
framework/src/iotbus/iotbus_gpio.c:423: ERROR: [IDT_M_TAB] please, no spaces at the start of a line

**Comment by seinfra** [comment]
Target : [d899aa2d2399c97834f21badf07c50de26c28e83] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [d899aa2d2399c97834f21badf07c50de26c28e83] - Code Rule Check OK.

File: framework/include/iotbus/iotbus_gpio.h [hunk 1]
```diff
@@ -181,6 +185,20 @@ int iotbus_gpio_register_cb(iotbus_gpio_context_h dev, iotbus_gpio_edge_e edge,
...
+ * @brief register interrupt callback.
+ *
+ * @details @b #include <iotbus/iotbus_gpio.h>
+ * @param[in] dev handle of gpio_context
+ * @param[in] edge gpio edge type
+ * @param[in] int_type interrupt type of gpio
+ * @param[in] enable enable/disable intterrupt
+ * @param[in] cb callback function of interrupt
+ * @return On success, 0 is returned. On failure, a negative value is returned.
+ * @since TizenRT v1.0
```
Comment (jeongchanKim) [inline framework/include/iotbus/iotbus_gpio.h]: Please check the version.

File: os/drivers/iotdev/iotbus_irq.c [hunk 2]
```diff
@@ -0,0 +1,316 @@
...
+		ret = sem_init(&dev->exclsem[i], 0, 1);
+		if (ret == -1) {
+			ibdbg("[%s] sem_init fail\n", iotbus_int_str[i]);
+			return;
+		}
+		dev->table[i] = NULL;
+	}
+
+	(void)register_driver(IOTBUS_IRQPATH, &iotbus_irq_ops, 0666, dev);
+}
```
Comment (jeongchanKim) [inline os/drivers/iotdev/iotbus_irq.c]: please remove unnecessary things.

File: os/include/signal.h [hunk 3]
```diff
@@ -215,6 +215,19 @@
 #define SIGEL_EVENT       CONFIG_SIG_SIGEL_EVENT
 #endif
 
+/* SIG_SIGEL_EVENT is used for event handling in Event Loop */
+#ifndef CONFIG_SIG_MESSAGING
```
Comment (jeongchanKim) [inline os/include/signal.h]: You'd better check this defines, because it is not for master branch.

File: os/drivers/iotdev/iotbus_irq.c [hunk 4]
```diff
@@ -0,0 +1,316 @@
...
+{
+	struct iotbus_upperhalf_s *dev = &g_iot_udev;
+
+	ibdbg("iotbus_irq_register()\n");
+
+	int ret, i;
+
+	g_irq_pid = kernel_thread(IOTBUS_IRQ_THREAD, IOTBUS_IRQ_PRIORITY, IOTBUS_IRQ_STACKSIZE, (main_t)iotbus_isr, (char *const *)NULL);
+	if (g_irq_pid > 0) {
+		ibdbg("Execute loading thread with pid %d\n", g_irq_pid);
```
Comment (jeongchanKim) [inline os/drivers/iotdev/iotbus_irq.c]: I think this dbg should be changed to ibvdbg

File: os/drivers/iotdev/iotbus_irq.c [hunk 5]
```diff
@@ -0,0 +1,316 @@
...
+	return;
+}
+
+void iotbus_irq_register(void)
+{
+	struct iotbus_upperhalf_s *dev = &g_iot_udev;
+
+	ibdbg("iotbus_irq_register()\n");
+
+	int ret, i;
```
Comment (jeongchanKim) [inline os/drivers/iotdev/iotbus_irq.c]: Please separate each declarations.

File: os/drivers/iotdev/iotbus_irq.c [hunk 6]
```diff
@@ -0,0 +1,316 @@
...
+	return;
+}
+
+void iotbus_interrupt_trigger(int irq, iotbus_int_type_e int_type)
+{
+	irqstate_t flags;
+	struct iotbus_upperhalf_s *dev = &g_iot_udev;
+	int ret;
+
+	ibdbg("Trigger Start![%d, %d]\n", irq, int_type);
```
Comment (jeongchanKim) [inline os/drivers/iotdev/iotbus_irq.c]: Please change this to vdbg

File: os/drivers/iotdev/iotbus_irq.c [hunk 7]
```diff
@@ -0,0 +1,316 @@
...
+{
+	int irq, int_type;
+	uint32_t int_val;
+	struct iotbus_upperhalf_s *dev = &g_iot_udev;
+
+	sigset_t set;
+	siginfo_t info;
+
+	sigemptyset(&set);
+	sigaddset(&set, SIG_IOTBUS_INT);
```
Comment (jeongchanKim) [inline os/drivers/iotdev/iotbus_irq.c]: In this setting means, you don't want to catch the SIG_IOTBUS_INT. Isn't it?

**Comment by seinfra** [comment]
Target : [70839673a7e3b3e74f5bb963ed14f7d16c1b6329] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [70839673a7e3b3e74f5bb963ed14f7d16c1b6329] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [599210b42fc36f22b9a08509484d8d6b4302d6bf] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [599210b42fc36f22b9a08509484d8d6b4302d6bf] - Code Rule Check OK.

**Comment by seinfra** [comment]
Target : [bd101c88d73650e37c41a52dad02e50a8159e32a] - Code Rule Check (C++) OK.

**Comment by seinfra** [comment]
Target : [bd101c88d73650e37c41a52dad02e50a8159e32a] - Code Rule Check OK.

File: os/drivers/iotdev/iotbus_irq.c [hunk 8]
```diff
@@ -0,0 +1,297 @@
+/****************************************************************************
```
Comment (Taejun-Kwon) [inline os/drivers/iotdev/iotbus_irq.c]: Please do not make driver for just operation, driver should control just one device.
If you want get message or callback, you have to expand each of driver. (Generally, tinyara using mq for it.)

File: framework/include/iotbus/iotbus_debug.h [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+ ****************************************************************************/
+#ifndef _IOTBUS_DEBUG_H__
+#define _IOTBUS_DEBUG_H__
+
+#include <tinyara/config.h>
+#include <debug.h>
+
+#ifdef CONFIG_CPP_HAVE_VARARGS
+#ifdef CONFIG_DEBUG_IOTBUS
+#define ibdbg(format, ...)			dbg(format, ##__VA_ARGS__)
```
Comment (sunghan-chang) [inline framework/include/iotbus/iotbus_debug.h]: All other definitions of debug are in debug.h. Do you need separated file only for iotbus?

File: framework/src/iotbus/iotapi_evt_handler.c [hunk 10]
```diff
@@ -212,19 +212,19 @@ int iotapi_handler_start(int fd)
  */
 void iotapi_initialize(void)
 {
-	idbg("[iotcom] init\n");
+	ibdbg("[iotcom] init\n");
 	int ret = 0, i = 0;
```
Comment (NilabhShant) [inline framework/src/iotbus/iotapi_evt_handler.c]: Please declare variable in new line.

File: os/drivers/iotdev/iotbus_irq.c [hunk 11]
```diff
@@ -0,0 +1,297 @@
...
+#include <unistd.h>
+#include <assert.h>
+#include <tinyara/fs/fs.h>
+#include <tinyara/kmalloc.h>
+#include <tinyara/iotbus_irq.h>
+#include <iotbus/iotbus_debug.h>
+
+#define IOTBUS_IRQ_THREAD		"IOTBUS_IRQ"
+#define IOTBUS_IRQ_PRIORITY		224
+#define IOTBUS_IRQ_STACKSIZE	2048
```
Comment (NilabhShant) [inline os/drivers/iotdev/iotbus_irq.c]: Please aline macro.

File: os/include/tinyara/iotbus_irq.h [hunk 12]
```diff
@@ -0,0 +1,90 @@
...
+/****************************************************************************
+ * Pre-processor Definitions
+ ****************************************************************************/
+
+#define IOTBUS_IRQPATH				"/dev/iotirq"
+#define IOTBUS_INT_TRIG(irq, type)	iotbus_interrupt_trigger(irq, type)
+
+#define IOTBUS_INTR_REGISTER		_IOTBUSIOC(1)
+#define IOTBUS_INTR_UNREGISTER		_IOTBUSIOC(2)
+#define IOTBUS_INTR_INFO			_IOTBUSIOC(3)
```
Comment (NilabhShant) [inline os/include/tinyara/iotbus_irq.h]: Please aline macro.

**Comment by an4967** [comment]
Logic is changed

**Pull request closed without merging**